src/util/result.h

```diff
@@ -31,54 +144,89 @@ struct Error {
+        }, std::forward<Args>(args)...);
+    }
 
-public:
-    Result(T obj) : m_variant{std::in_place_index_t<1>{}, std::move(obj)} {}
-    Result(Error error) : m_variant{std::in_place_index_t<0>{}, std::move(error.message)} {}
+    template <typename Fn, typename... Args>
+    void Construct(const Fn& fn, Warning warning, Args&&... args)
+    {
+        this->AddWarning(std::move(warning.message));
+        Construct(fn, std::forward<Args>(args)...);
+    }
 
-    //! std::optional methods, so functions returning optional<T> can change to
-    //! return Result<T> with minimal changes to existing code, and vice versa.
-    bool has_value() const noexcept { return m_variant.index() == 1; }
-    const T& value() const LIFETIMEBOUND
+    template <typename Fn, typename OT, typename OF, typename... Args>
+    void Construct(const Fn& fn, Result<OT, OF>&& other, Args&&... args)
     {
-        assert(has_value());
-        return std::get<1>(m_variant);
+        *this << std::move(other);
+        Construct(fn, std::forward<Args>(args)...);
     }
-    T& value() LIFETIMEBOUND
+
+    template <typename OT, typename OF>
+    void MoveConstruct(Result<OT, OF>& other)
     {
-        assert(has_value());
-        return std::get<1>(m_variant);
+        *this << std::move(other);
+        if (other) this->MoveValue(other); else if (other.m_info) this->m_info->failure = std::move(other.m_info->failure);
     }
-    template <class U>
-    T value_or(U&& default_value) const&
+
+    template <typename FT, typename FF>
+    friend class Result;
+
+public:
+    //! Constructors, destructor, and assignment operator.
+    template <typename... Args>
+    Result(Args&&... args)
     {
-        return has_value() ? value() : std::forward<U>(default_value);
+        Construct([this](auto&&... x) { this->ConstructValue(std::forward<decltype(x)>(x)...); }, std::forward<Args>(args)...);
     }
-    template <class U>
-    T value_or(U&& default_value) &&
+    template <typename OT, typename OF>
+    Result(Result<OT, OF>&& other) { MoveConstruct(other); }
+    Result& operator=(Result&& other)
     {
-        return has_value() ? std::move(value()) : std::forward<U>(default_value);
+        if (*this) this->DestroyValue(); else this->m_info->failure.reset();
+        MoveConstruct(other);
+        return *this;
+    }
+    ~Result() { if (*this) this->DestroyValue(); }
+
+    //! Operator moving warning and error messages from one result to another.
+    //! Only moves message strings, does not change success or failure values of
+    //! either Result object.
+    template <typename OT, typename OF>
+    Result<OT, OF>&& operator<<(Result<OT, OF>&& other)
+    {
+        if (other.m_info && !this->m_info) {
+            this->m_info.reset(new detail::ErrorInfo<F>{.errors = std::move(other.m_info->errors),
+                                                        .warnings = std::move(other.m_info->warnings)});
```

Comment:
Thank you for the thoughtful response, that was very helpful. Your usage of move semantics here now makes sense to me.

Now I'm just confused that the [designated initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers) compiles fine even though the spec says it's since C++20 and I haven't configured with `--enable-c++20` (as does the CI I think)?

Reply:
> Now I'm just confused that the [designated initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers) compiles fine even though the spec says it's since C++20 and I haven't configured with `--enable-c++20` (as does the CI I think)?

Oh, we are just outside of the c++17 spec there. The build enables designated initializers as an extension since #24531